### PR TITLE
Refactoring factory/tests

### PR DESCRIPTION
**Pull request by iimpulse**
@kingmanzhang  Take a look at some of my changes. The service test really doesn't need the top part because I am mocking in the test.. However if you were to use the entire file thats what you would need to do. Implementing ServiceUnitTest  gives you a mock to the service in "service" everywhere by default no need to use @Shared . 

Refactoring the Loinc2HpoFactorySpec should be more straight forward.. you have 2 choices mocks ore the entire file ( the latter is more brittle to change in the annotations ). Lets discuss my changes so you understand.

@pnrobinson HpoLoincServiceSpec.groovy , mocking in spock. This is only a singular object. You can do more adanced mocking like every other call we want to return this vs that.

### CODE REVIEW

File: grails-app/services/hpo/api/HpoLoincService.groovy [hunk 1]
```diff
@@ -13,16 +14,16 @@ import org.monarchinitiative.phenol.ontology.data.TermId
...
 
   Set<LoincEntry> searchByHpo(TermId query) {
-    Set<LoincId> loincIds = loinc2HpoFactory.reverseAnnotationMap().get(query)
-    Map<LoincId, LoincEntry> loincEntryMap = loinc2HpoFactory.loincEntryMap()
+    Set<LoincId> loincIds = hpoLoinc.getReverseAnnotationMap().get(query)
+    Map<LoincId, LoincEntry> loincEntryMap = hpoLoinc.getLoincEntryMap()
 
     Set<LoincEntry> result = new HashSet<>()
 
-    for (LoincId loincid : loincIds) {
```
Comment (iimpulse) [inline grails-app/services/hpo/api/HpoLoincService.groovy]: Get Groovy about it. Easy Closure pattern here. Note.. you cant use "break" inside the closure.

File: src/main/groovy/hpo/api/model/LoincTerm.groovy [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package hpo.api.model
+
+/**
+ * A minimum representation of a LOINC term
+ */
+class LoincTerm {
```
Comment (iimpulse) [inline src/main/groovy/hpo/api/model/LoincTerm.groovy]: This is a groovy class. They automatically have setters/getters attached based on the member variables no need to create them.

File: src/test/groovy/hpo/api/HpoLoincServiceSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  void "search loinc by hpo"() {
+
+    given:
+      TermId query = TermId.of("HP:0004363")
+      Map<LoincId, LoincEntry> mockLoincEntryMap = new HashMap<>()
+      Map<TermId, Set<LoincId>> mockReverseMap = new HashMap<>()
+
+      mockLoincEntryMap.put(buildMockLoincId("38230-9"), buildMockLoincEntry(getOneLoincLine()))
+      mockReverseMap.put(TermId.of("HP:0004363"), new HashSet<LoincId>(Arrays.asList(new LoincId("38230-9"))))
+      service.hpoLoinc = Mock(Loinc2Hpo)
```
Comment (iimpulse) [inline src/test/groovy/hpo/api/HpoLoincServiceSpec.groovy]: mocking the entire injected bean for the service

File: src/test/groovy/hpo/api/HpoLoincServiceSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      service.hpoLoinc.getReverseAnnotationMap() >> mockReverseMap
+      service.hpoLoinc.getLoincEntryMap() >> mockLoincEntryMap
```
Comment (iimpulse) [inline src/test/groovy/hpo/api/HpoLoincServiceSpec.groovy]: when method gets called return the mock

File: grails-app/conf/spring/resources.groovy [hunk 5]
```diff
@@ -11,6 +12,9 @@ beans = {
     hpoAssociationFactory(HpoAssociationFactory, hpoOntology)
     hpoAssociation(hpoAssociationFactory: "getInstance")
 
+    hpoLoincFactory(Loinc2HpoFactory)
```
Comment (iimpulse) [inline grails-app/conf/spring/resources.groovy]: Create a loinc factory bean. Then create the hpoLoinc injectable bean with our data by calling the previous beans getinstance method which returns a Loinc2Hpo Instance... all only once on startup of the app.

Comment (kingmanzhang) [reply grails-app/conf/spring/resources.groovy]: This is nice!

File: src/test/groovy/hpo/api/HpoLoincServiceSpec.groovy (hunk 4 above)
Comment (kingmanzhang) [reply src/test/groovy/hpo/api/HpoLoincServiceSpec.groovy]: It seems setupSpec() tries to read all annotation and loincEntries anyway. Should we also mock Loinc2Hpo?

Comment (iimpulse) [reply src/test/groovy/hpo/api/HpoLoincServiceSpec.groovy]: This is true. On my first pass I didn't look at the code you have commented out. So I just implemented the whole thing.. Then I used a test level mock here. The reading of all annotation files part can probably be deleted. I was just doing it to show how you would do it.

Comment (kingmanzhang) [reply src/test/groovy/hpo/api/HpoLoincServiceSpec.groovy]: I deleted it now.

File: src/test/groovy/hpo/api/util/Loinc2HpoFactorySpec.groovy [hunk 6]
```diff
@@ -18,32 +20,36 @@ class Loinc2HpoFactorySpec extends Specification {
...
+      resultMap.keySet().toString() == expectedKeys
+
+    where:
+    expectedSize | expectedKeys
+    3            | '[38230-9, 738-5, 5778-6]'
+
   }
 
-  def "ReverseAnnotationMap"() {
+  def "ReverseAnnotationMap returns #desc"() {
```
Comment (iimpulse) [inline src/test/groovy/hpo/api/util/Loinc2HpoFactorySpec.groovy]: take a look at my changes here.. @kingmanzhang .. can you refactor the last test to be more spock like?

Comment (kingmanzhang) [reply src/test/groovy/hpo/api/util/Loinc2HpoFactorySpec.groovy]: This is really cool feature! Thanks for showing me this.

**Pull request merged**